futures-util/src/lock/mutex.rs

```diff
@@ -0,0 +1,246 @@
+            .field("was_acquired", &self.mutex.is_none())
+            .field("mutex", &self.mutex)
+            .field("wait_key", &(
+                    if self.wait_key == WAIT_KEY_NONE {
+                        None
+                    } else {
+                        Some(self.wait_key)
+                    }
+                ))
+            .finish()
+    }
+}
+
+impl<T> FusedFuture for MutexAcquire<'_, T> {
+    fn is_terminated(&self) -> bool {
+        self.mutex.is_none()
+    }
+}
+
+impl<'a, T> Future for MutexAcquire<'a, T> {
+    type Output = MutexGuard<'a, T>;
+
+    fn poll(mut self: Pin<&mut Self>, lw: &LocalWaker) -> Poll<Self::Output> {
+        let mutex = self.mutex.expect("polled MutexAcquire after completion");
+
+        if let Some(lock) = mutex.try_lock() {
+            mutex.remove_waker(self.wait_key, false);
+            self.mutex = None;
+            return Poll::Ready(lock);
+        }
+
+        {
+            let mut waiters = mutex.waiters.lock().unwrap();
+            if self.wait_key == WAIT_KEY_NONE {
+                self.wait_key = waiters.insert(Waiter::Waiting(lw.clone().into_waker()));
+                if waiters.len() == 1 {
+                    mutex.state.fetch_or(HAS_WAITERS, Ordering::Relaxed); // released by mutex unlock
+                }
+            } else {
+                waiters[self.wait_key].register(lw);
+            }
+        }
+
+        // Ensure that we haven't raced `MutexGuard::drop`'s unlock path by
+        // attempting to acquire the lock again.
+        if let Some(lock) = mutex.try_lock() {
+            mutex.remove_waker(self.wait_key, false);
+            self.mutex = None;
+            return Poll::Ready(lock);
+        }
+
+        Poll::Pending
+    }
+}
+
+impl<T> Drop for MutexAcquire<'_, T> {
+    fn drop(&mut self) {
+        if let Some(mutex) = self.mutex {
+            // This future was dropped before it acquired the mutex.
+            //
+            // Remove ourselves from the map, waking up another waiter if we
+            // had been awoken to acquire the lock.
+            mutex.remove_waker(self.wait_key, true);
+        }
+    }
+}
+
+/// An RAII guard returned by the `lock` and `try_lock` methods.
+/// When this structure is dropped (falls out of scope), the lock will be
+/// unlocked.
+pub struct MutexGuard<'a, T: 'a> {
+    mutex: &'a Mutex<T>,
+}
+
+impl<T: fmt::Debug> fmt::Debug for MutexGuard<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("MutexGuard")
+            .field("value", &*self)
+            .field("mutex", &self.mutex)
+            .finish()
+    }
+}
+
+impl<T> Drop for MutexGuard<'_, T> {
+    fn drop(&mut self) {
+        let old_state = self.mutex.state.fetch_and(!IS_LOCKED, Ordering::AcqRel);
+        if (old_state & HAS_WAITERS) != 0 {
+            let mut waiters = self.mutex.waiters.lock().unwrap();
+            if let Some((_i, waiter)) = waiters.iter_mut().next() {
```

Comment:
This approach is quite unfair, but we don't really have a better way to do this with `Slab`. We could perhaps use a `VecDeque` with elements that get nulled out instead as well as an offset number that is incremented each time the queue is popped (so that waiters can find themselves and update their entries appropriately). That would give approximately FIFO order, but is a bit more complex. WDYT?

Reply:
That approach would also make it possible for a single long-waiting task that didn't advance its future to cause the wait set to grow indefinitely.

Reply:
Another option is to have `next: usize` and `prev: usize` fields in `Waiter` (but really, `u32` or even `u16` should be enough), which indicate the next and previous waiter in the linked list. Then we can build a FIFO doubly linked list.

Reply:
So this wakes only one waiter instead of all of them, to avoid a thundering herd? That seems reasonable (apart from the fairness question).
One question that I had in my mind when I thought about implementing this is whether this release should already mark the mutex as locked again for the other waiter, even though that one doesn't yet know about it. That would avoid the issue that another call to lock can acquire the mutex in between.
But I'm not sure if it introduces other issues, since that `MutexAcquire` would now have acquired the lock without a direct call on it. However that passively acquired lock could still be released when the `MutexAcquire ` gets dropped without another `poll` call onto it which yields the `MutexGuard`.

Reply:
@Matthias247 That could possibly increase fairness, but it reduces throughput since it doesn't necessarily allow the next task to poll the lock to acquire it, and it makes it harder to recover from the case where a task is cancelled and has to figure out whether or not to pass the lock along to another waiter.

Reply:
I don't think the cancellation scenario gets a lot tricker, since the future already needs to track whether it got signalled and needs to wake another waiter if that's the case during `drop`.

However it might really introduce some new issues (e.g. deadlocks) during cancellation, e.g. if the future inside a task is not immediately destructed when done but hold alive due to a still waker references that are stored somewhere. As far as I understand some executors might behave that way (even though it should certainly be discouraged).

Therefore I agree that the strategy makes sense.